### ISSUE TITLE
More trouble with OOPS

### ISSUE THREAD

**Issue by DavidGriffith** [DavidGriffith/inform6lib#11]
There might be some new trouble with the OOPS verb.  Playing around with Uninvited, I found this:

    >PUT GEM IN NII
    You see no nii here.

    >OOPS NICHE
    You see no nii here.

However, misspelling the first noun and then doing OOPS works correctly.

**Comment by bignose-debian** [comment]
This only exhibits when the word to be corrected is at particular positions.

Correcting the first noun (the one from `noun`) works fine:

```
> search nii
You see no nii here.

> oops niche
In the niche is a dusty journal
```

When attempting to correct the word in `second`, the bug is demonstrated:
```
> put matchbox in nii
You see no nii here

> oops nii
You see no nii here.
```

Also when the word to correct is the verb:
```
> thrrr knife
That's not a verb I recognise.

> oops throw
That's not a verb I recognise.

> throw knife
What do you want to throw the chef knife at?
```

**Comment by bignose-debian** [comment]
I've tried with the `demos/toyshop.inf` story file (in my case, compiled to Z-code version 5).

```
[Replaying commands.]

>get string
You take the balloon by its string. It's buoyant!

>put ball in car
You can't see any such thing.

>oops balloon
You put the helium balloon into the little red car.

>take balloon
You take the balloon by its string. It's buoyant!

>put balloon in truck
You can't see any such thing.

>oops car
You put the helium balloon into the little red car.

>vault car
That's not a verb I recognise.

>oops enter
Sorry, that can't be corrected.

>get balloon
You take the balloon by its string. It's buoyant!

>put balloon under car
I didn't understand that sentence.

>oops in
I only understood you as far as wanting to in.

>put balloon xyzzy car
I didn't understand that sentence.

>oops in
I only understood you as far as wanting to in.

>quit
Are you sure you want to quit? yes
```

What is the expected behaviour for the OOPS command? Which of the above are not behaving as expected?

**Comment by bignose-debian** [comment]
This is the Toyshop command record I used for the above.
[toyshop.rec.txt](https://github.com/DavidGriffith/inform6lib/files/222889/toyshop.rec.txt)

**Comment by bignose-debian** [comment]
* For an unknown verb corrected by OOPS:
  * Uninvited responds as though the command was just another unknown verb: the same message “That's not a verb I recognise.” is returned. This makes it unclear whether the verb OOPS is even recognised.
  * Toyshop responds with “Sorry, that can't be corrected.”, which I think is better. I don't know why the behaviour is different though.

* For an unknown first noun (the `noun` value) corrected by OOPS:
  * Uninvited and Toyshop both respond as though the corrected command was typed. This is IMO the desirable behaviour.

* For an unknown second noun (the `second` value) corrected by OOPS:
  * Uninvited responds as though the same command (with the incorrect `second`) was typed again: “You see no nii here.”.
  * Toyshop responds by performing the command with the `second` value corrected. This is ideal, I think.

**Comment by vlaviano** [comment]
Having read through this issue, I have a question:

The version of Uninvited in the archive was compiled in 2012. Is this the version that you guys are talking about? If so, I'm not sure how its behavior is relevant to the current state of OOPS.

I compiled the test program for issue #30 against 6.12.1pre today and ran a few tests. I see the same behavior correcting verbs, nouns, prepositions, and second nouns that @bignose-debian reports for Toyshop. Here's my assessment of this behavior:

Verbs: "Sorry, that can't be corrected" is better than "That's not a verb I recognise", but even better would be to actually correct the verb. We should look at changing this. It might be as simple as a well-placed "oops_from = verb_wordnum;" when etype == VERB_PE.

Nouns: Ok.

Prepositions: Broken. See the PUT BALLOON XYZZY CAR example above.

Second nouns: Generally ok, but I did notice this:
```
Somewhere
You're not sure where you are.

You can see four rocks, a stone and a girl here.

>give stone to girl
(first taking the stone)
The girl doesn't seem interested.

>give stone to grl
You can't see any such thing.

>oops girl
You can only do that to something animate.

>show stone to girl
The girl is unimpressed.

>show stone to grl
You can't see any such thing.

>oops girl
The girl is unimpressed.
```

**Comment by DavidGriffith** [comment]
I think we can safely use just #30's test program.  Uninvited is just where I first noticed the problem.

**Comment by DavidGriffith** [comment]
Working on the second noun problem.  With trace set to 2, we run the above tests again.  First the SHOW sequence because it produces the correct response.
```
>TRACE 2
[Parser tracing set to level 2.]

>SHOW STONE TO GIRL
[ "show" show / "stone" stone / "to" to / "girl" girl ]
The girl is unimpressed.

>SHOW STONE TO FOO
[ "show" show / "stone" stone / "to" to / "foo" ? ]

>OOPS GIRL
[ "show" show / "stone" stone / "to" to / "girl" girl ]
```
For the first line of input, we see that all tokens match up with the names of objects.  For the second line, the "foo" token doesn't match up and so we see a question mark.  After the OOPS command is given, we see that the tokenized line sent to the parser is identical to the first line of input.  This is as it should be.  Now look at how GIVE happens:

```
>GIVE STONE TO GIRL
[ "give" give / "stone" stone / "to" to / "girl" girl ]

>GIVE STONE TO FOO
[ "give" give / "stone" stone / "to" to / "foo" ? ]

>OOPS GIRL
[ "give" give / "stone" stone / "girl" girl / "foo" ? ]
```
Notice how the tokenization is fouled up in the third line of the GIVE sequence.

**Comment by DavidGriffith** [comment]
From grammar.h:
```
Verb 'give' 'feed' 'offer' 'pay'
    * held 'to' creature                        -> Give
    * creature held                             -> Give reverse
    * 'over' held 'to' creature                 -> Give;
...
Verb 'show' 'display' 'present'
    * creature held                             -> Show reverse
    * held 'to' creature                        -> Show;
```
Notice that for these two `* creature held` and `* held 'to' creature` are in opposite orders.  I reordered the `give` verb like this:
```
 Verb 'give' 'feed' 'offer' 'pay'
    * creature held                             -> Give reverse
    * held 'to' creature                        -> Give
    * 'over' held 'to' creature                 -> Give;
```
This seems to fix the second noun problem.  I've also checked to make sure both forms work -- `GIVE GRRL STONE` as well as `GIVE STONE TO GRRL` followed by `OOPS GIRL` works.

Why did this work?

**Comment by DavidGriffith** [comment]
This partial fix is the best we can do for now.  A more correct fix will have to wait for 6.12.2.  I'm going to close this issue now and open a new one that better describes the problem.
